api/internal/git/repospec.go

```diff
@@ -208,23 +233,24 @@ func parseHostSpec(n string) (string, string) {
 		}
```

Comment:
This logic of `parseHostSpec` is really confusing, and since we are further complicating the code path below and adding a new `parseDomain`, I think we should rewrite a good portion of `parseHostSpec` to be less convoluted. 

For example, we can by first handling the special cases of `gh:` and `git@`. Then, we can parse out the schemes of the URL. Then get the host name. (Maybe that won't work for whatever reason, in which case feel free to simplify in a different way or at least add comments explaining what's happening and why it is necessary). At minimum we need more comments - `parseDomain` and lines 236-245 are hard to reason about. I'm constantly trying to figure out what is `n` now and what part of the URL are we trying to extract? 

Lines 128-147 have a similar problem, but since this PR didn't add or change that code, we don't need to clean it up in this PR.

Reply:
I hesitate to re-write `parseHostSpec` now because it currently fails to deal with some edge cases and I don't have the confidence to rewrite it in this PR without introducing more bugs.

However, I have refactored `parseGitURL` and commented `parseGitURL` (https://github.com/kubernetes-sigs/kustomize/pull/4652#discussion_r936926790), `parseHostSpec`, and `parseDomain` more heavily. Lmk if anything is still hard to read.

Reply:
A quick question in case I miss some context. Is there a specific reason to not use an existing library (e.g. [go-git](https://github.com/go-git/go-git)) here but implement our own?

Reply:
No idea on my end. I assumed we had a good reason for implementing our own, so I stuck with it. @natasha41575?

Reply:
I don't have the context either. The main reasons I would be hesitant to rewrite this with go-git:

1. This would be a very big refactor, and is low priority for kustomize. 
2. We don't have very good tests for this code unfortunately, so a big refactor is risky. 
3. We will have to check with kubectl before we add a go dependency. They sometimes are picky about adding dependencies to the kubernetes/kubernetes repo.